### PR TITLE
circleci: publish test reports

### PR DESCRIPTION
**Pull request by alxndrsn**

### CODE REVIEW

**Comment by matthew-white** [comment]
I haven't looked deeply into this, but the idea of seeing test insights in CircleCI seems potentially useful!

**Comment by alxndrsn** [comment]
Example:

* working: https://app.circleci.com/insights/github/alxndrsn/odk-central-backend/workflows/workflow/tests?branch=circle-ci-test-reports
* current master (not working): https://app.circleci.com/insights/github/getodk/central-backend/workflows/workflow/tests?branch=master

**Comment by alxndrsn** [comment]
I've rebased this onto latest `master`.  Some additional thoughts:

* the current test output location seems esoteric (`~/junit/`)
* it's a shame that we don't get the standard text output _as well as_ the JUnit XML output

**Comment by matthew-white** [comment]
> the current test output location seems esoteric (`~/junit/`)

That seems OK to me. 🤷

> it's a shame that we don't get the standard text output _as well as_ the JUnit XML output

I do think that that has the potential to make things harder to debug. With the current output, you can scroll up to a failed test to check whether there was any logging or an unhandled promise rejection before the test failure. But with the JUnit output, it looks like you just see the test failure and its associated stack trace. If there's a reasonable way to output both, that seems ideal. It's unfortunate though that Mocha doesn't support that out of the box. 😭

File: .circleci/config.yml [hunk 1]
```diff
@@ -14,5 +14,7 @@ jobs:
       - run: make check-file-headers
       - run: npm ci --legacy-peer-deps
       - run: node lib/bin/create-docker-databases.js
-      - run: make test-full
-
+      - run: mkdir ~/junit
+      - run: MOCHA_FILE=~/junit/test-results.xml make test-ci
```
Comment (sadiqkhoja) [inline .circleci/config.yml:18]: I am interesting in the recent feature of CircleCI "rerunning failed tests", for that we would need this step to be:

```
      - run: 
          name: Run tests
          command: |
            mkdir test-results
            TEST_FILES=$(circleci tests glob "test/**/*.js")
            echo "$TEST_FILES" | circleci tests run --command="xargs npx mocha --exit --reporter mocha-junit-reporter --reporter-options mochaFile=./test-results/junit.xml"
      - store_test_results:
          path: test-results
```

**Comment by sadiqkhoja** [comment]
How about using [mocha-multi-reporters](https://www.npmjs.com/package/mocha-multi-reporters) with mocha-junit-reporter?

File: .circleci/config.yml (hunk 1 above)
Comment (alxndrsn) [reply .circleci/config.yml]: :+1: that could be a good next step

**Comment by alxndrsn** [comment]
> How about using [mocha-multi-reporters](https://www.npmjs.com/package/mocha-multi-reporters) with mocha-junit-reporter? I am happy without text output as well

I've implemented this directly - it turns out it's quite simple:

```
const mocha = require('mocha');
const MochaJunitReporter = require('mocha-junit-reporter');

module.exports = function CiMochaReporter(runner) {
  new mocha.reporters.Spec(runner);
  new MochaJunitReporter(runner, { mochaFile: './junit-reports/test-results.xml' });
};
```

**Comment by alxndrsn** [comment]
@matthew-white @sadiqkhoja this is ready for review again.

* test results are now stored in `./junit-reports/`
* normal text output **and** junit XML output are both included

Example failure at:
* logs: https://app.circleci.com/pipelines/gh/alxndrsn/odk-central-backend/1052/workflows/98e85503-79a0-4699-aaaa-875d493a9f03/jobs/976/steps
* CircleCI test integration: https://app.circleci.com/pipelines/gh/alxndrsn/odk-central-backend/1052/workflows/98e85503-79a0-4699-aaaa-875d493a9f03/jobs/976/tests

**Review by sadiqkhoja — APPROVED** [review]
Looks great to me

**Pull request merged**